apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/service/NamespaceService.java

```diff
@@ -165,20 +173,29 @@ public List<NamespaceBO> findNamespaceBOs(String appId, Env env, String clusterN
       throw new BadRequestException("namespaces not exist");
     }
 
-    List<NamespaceBO> namespaceBOs = new LinkedList<>();
+    List<NamespaceBO> namespaceBOs = new CopyOnWriteArrayList<>();
+    CountDownLatch latch = new CountDownLatch(namespaces.size());
     for (NamespaceDTO namespace : namespaces) {
+      executorService.submit(() -> {
+        NamespaceBO namespaceBO;
+        try {
+          namespaceBO = transformNamespace2BO(env, namespace);
+          namespaceBOs.add(namespaceBO);
+        } catch (Exception e) {
+          LOGGER.error("parse namespace error. app id:{}, env:{}, clusterName:{}, namespace:{}",
+              appId, env, clusterName, namespace.getNamespaceName(), e);
+          throw e;
+        } finally {
+          latch.countDown();
+        }
+      });
 
-      NamespaceBO namespaceBO;
-      try {
-        namespaceBO = transformNamespace2BO(env, namespace);
-        namespaceBOs.add(namespaceBO);
-      } catch (Exception e) {
-        LOGGER.error("parse namespace error. app id:{}, env:{}, clusterName:{}, namespace:{}",
-            appId, env, clusterName, namespace.getNamespaceName(), e);
-        throw e;
-      }
     }
-
+    try {
+      latch.await();
```

Comment:
Is it better to use the method with timeout here? e.g. CountDownLatch.await(long timeout, TimeUnit unit)

Reply:
I don't feel it's necessary

Reply:
I mainly consider two points:

-  The remote interactions involved are all internal systems, and there is no unpredictable long blocking

- If you use  `CountDownLatch.await(long timeout, TimeUnit unit)` , Not very good to evaluate this timeout parameter, which will complicate the problem

Reply:
Yes, the timeout is really not very good to set.